### ISSUE TITLE
vfs/refresh does not cache entire directory

### ISSUE THREAD

**Issue by H1XYZ** [ncw/rclone#2946]
Hello,
Facing a strange issue with rclone VFS.

Version I am running: `v1.45-133-g291f2709-beta` (although this happened on stable as well)

It seems like the mount is not seeing a large chunk of my directories:

> rclone mount 323 dirs
> rclone vfs/refresh = 88 dirs
> rclone vfs/forget = 323 dirs

Commands:

> ls -lrth /mnt/remote/Media/TV |wc -l
> 88
>
> rclone rc vfs/forget dir=Media/TV
> {
>     "forgotten": [
>         "Media/TV"
>     ]
> }
>
> ls -lrth /mnt/remote/Media/TV |wc -l
> 323
> 

This happens across any directory, here is another example:

> ls -lrth /mnt/remote/Media/Movies/HD |wc -l
> 435
> 
> rclone rc vfs/forget dir=Media/Movies/HD
> {
> 	"forgotten": [
> 		"Media/Movies/HD"
> 	]
> }
> 
> ls -lrth /mnt/remote/Media/Movies/HD |wc -l
> 628

Any thoughts would be appreciated

**Comment by H1XYZ** [comment]
Further testing seems to reveal that this may be related to ` --fast-list` on mount command. 
It maybe causing the issue. Requires further testing by others.

**Comment by lcasey001** [comment]
> If the parameter recursive=true is given the whole directory tree will get refreshed. This refresh will use --fast-list if enabled.

Does that mean its being used internally when `recursive=true` is enabled  and no need to add it to the `rc` command ? Or do we have to add it to the `vfs/refresh` command for it to kick in?

Also what benefit would running `vfs/refresh` give you vs running something like `find {mount}` ?

**Comment by EnorMOZ** [comment]
> If the parameter recursive=true is given the whole directory tree will get refreshed. This refresh will use --fast-list if enabled.

Does that mean its being used internally when `recursive=true` is enabled  and no need to add it to the `rc` command ? Or do we have to add it to the `vfs/refresh` command for it to kick in?

Also what benefit would running `vfs/refresh` give you vs running something like `find {mount}` ?

**Comment by ncw** [comment]
> Does that mean that --fast-list is being used by default when recursive=true is enabled and no need to add it to the rc command ? Or do we have to add specify it explicitly (eg vfs/refresh)?

That is correct, `--fast-list` should be automatically enabled for `vfs/refresh` if the backend supports it.

**Comment by H1XYZ** [comment]
@ncw any insights or ideas on this other issue?

**Comment by ncw** [comment]
> Further testing seems to reveal that this may be related to ` --fast-list` on mount command.
> It maybe causing the issue. `--fast-list` to rc command has no affect

Are you saying that the issue went away if you stopped using `--fast-list` on the mount?

As far as I can see `--fast-list` should do nothing when using the mount.

If the issue is still present a bit more info would be helpful.

Can you show the command line you are using for the mount?

Which backend are you using?

If you could do diff the output of the `ls -lrth /mnt/remote/Media/TV` commands and work out what is missing that would be helpful too.

> Also what benefit would running vfs/refresh give you vs running something like find {mount} ?

    rclone rc vfs/refresh recursive=true

should be much quicker than using find

**Comment by EnorMOZ** [comment]
I was helping him troubleshoot this in the cloudbox discord channel.  

> Are you saying that the issue went away if you stopped using --fast-list on the mount?

In @H1XYZ case yes. In my case it actually caused vfs/refresh to take forever and not finish before the timeout. Once added back to the mount it worked within 2minutes and no missing dirs.
> Can you show the command line you are using for the mount?


rclone mount command used: `/usr/bin/rclone mount --config=/home/user/.config/rclone/rclone.conf --allow-other --rc --rc-addr=localhost:5572 --fast-list --drive-skip-gdocs --vfs-read-chunk-size=64M --vfs-read-chunk-size-limit=2048M --buffer-size=64M --max-read-ahead=256M --poll-interval=30s --dir-cache-time=168h --timeout=10m --transfers=20 --checkers=16 --fuse-flag=sync_read --fuse-flag=auto_cache --drive-chunk-size=64M --umask=002 --syslog -v google: /mnt/remote`

rclone vfs refresh used : `/usr/bin/rclone --config /home/user/.config/rclone/rclone.conf --timeout=1h -vvv rc --rc-addr=localhost:5572 vfs/refresh dir=Media recursive=true`

> Which backend are you using?

gdrive

I have not been able to reproduce his issue. Having `--fist-list` on the mount command though does seen to cause  rclone vfs/refresh command to run a lot faster in my and other cases. In my case it finishes up a 225TB gdrive in 1.5minutes. However once I remove it from the mount command it caused it to be very slow where it would time out. 

There seems to be two issues here. 
1) vfs/refresh causing dirs to not be cached or are missing. (H1XYZ) issue
2) Removing vfs/refresh from the mount command causes it to not run properly and/or be very slow and timed out. (Others have experienced this)

**Comment by ncw** [comment]
> There seems to be two issues here.
> 
> 1. vfs/refresh causing dirs to not be cached or are missing. (H1XYZ) issue
> 2. Removing --fast-list from the mount command causes vfs/refresh to not run properly and/or be very slow and timed out. (Others have experienced this)

The second issue is a bug which I've attempted to fix here - can you have a go?

https://beta.rclone.org/branch/v1.45-168-g06c1447e-fix-2946-dirtree-beta/ (uploaded in 15-30 mins)

The first is a bug too, but I haven't managed to replicate it either.

**Comment by H1XYZ** [comment]
@ncw 
pre vfs / refresh: https://pastebin.com/raw/XBRh7Bz4 (323 count)
post vfs / refresh: https://pastebin.com/raw/zfMDXyaV (122 count)

I have done this multiple times, the post refresh results vary each time in terms of total dir count, which tells me it stalls randomly.

**Comment by EnorMOZ** [comment]
> The second issue is a bug which I've attempted to fix here - can you have a go?
https://beta.rclone.org/branch/v1.45-168-g06c1447e-fix-2946-dirtree-beta/ (uploaded in 15-30 mins)

I can confirm this does fix the issue of slow running vfs/refresh when mount does not have --fast-list.

**Comment by EnorMOZ** [comment]
Is there a way to bump up the logs on vfs/refresh or something see why it might not be caching all of @H1XYZ  dirs ?

**Comment by ncw** [comment]
> > The second issue is a bug which I've attempted to fix here - can you have a go?
> > https://beta.rclone.org/branch/v1.45-168-g06c1447e-fix-2946-dirtree-beta/ (uploaded in 15-30 mins)
> 
> I can confirm this does fix the issue of slow running vfs/refresh when mount does not have --fast-list.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.46



> Is there a way to bump up the logs on vfs/refresh or something see why it might not be caching all of @H1XYZ dirs ?

You could stick some `log.Printf` in the source - that is probably what I'd do!

Can you try an experiment.  I suspect it is the --fast-list support for drive which is the problem.

Can you do

    rclone lsf -R gdrive:troublesomedir | sort > test0
    rclone lsf --fast-list -R gdrive:troublesomedir | sort > test1
    rclone lsf --fast-list -R gdrive:troublesomedir | sort > test2

And then run

    diff -u test0 test1
    diff -u test1 test2

And see what differences there are - if any?

If --fast-list is working properly there should be no differences.

**Comment by ncw** [comment]
I should have thought of this before, but if there are duplicate directories this could explain everything!  So run `rclone dedupe`...

**Comment by H1XYZ** [comment]
@ncw 
i did run dedupe - there are no duplicates. The issue also happens on other directories so i know it's not that.

The diff commands had no difference noted. 

Also did:
`rclone lsf -R google:Media/TV/ | wc -l
51811
rclone lsf --fast-list -R google:Media/TV/ | wc -l
51811`

There was no difference in files or count however --fast-list was significantly faster when running

**Comment by H1XYZ** [comment]
Mount is not in use when I did this. I stopped all things accessing it when I did the tests

**Comment by ncw** [comment]
> Mount is not in use when I did this. I stopped all things accessing it when I did the tests

OK.

Can I see a complete log with `-vv` of you running `vfs/refresh` ?

**Comment by H1XYZ** [comment]
we checked that as well, one of the first few things but it had nothing in particular pointing to this. Everything looked normal in the log...

**Comment by ncw** [comment]
> we checked that as well, one of the first few things but it had nothing in particular pointing to this. Everything looked normal in the log...

I'd still like to see a log if possible :-)

**Comment by H1XYZ** [comment]
> > we checked that as well, one of the first few things but it had nothing in particular pointing to this. Everything looked normal in the log...
> 
> I'd still like to see a log if possible :-)

command run:

> /usr/bin/rclone \
>   --config /home/seed/.config/rclone/rclone.conf \
>   --timeout=1h \
>   -vvvvv \
>   --rc-addr=localhost:5572 \
>   rc vfs/refresh dir=Media recursive=true

Output:

> 2019/02/06 11:55:50 DEBUG : rclone: Version "v1.45-171-gbdc63853-beta" starting with parameters ["/usr/bin/rclone" "--config" "/home/seed/.config/rclone/rclone.conf" "--timeout=1h" "-vvvvv" "--rc-addr=localhost:5572" "rc" "vfs/refresh" "dir=Media" "recursive=true"]
> {
> 	"result": {
> 		"Media": "OK"
> 	}
> }
> 2019/02/06 11:59:32 DEBUG : 4 go routines active
> 2019/02/06 11:59:32 DEBUG : rclone: Version "v1.45-171-gbdc63853-beta" finishing with parameters ["/usr/bin/rclone" "--config" "/home/seed/.config/rclone/rclone.conf" "--timeout=1h" "-vvvvv" "--rc-addr=localhost:5572" "rc" "vfs/refresh" "dir=Media" "recursive=true"]

Directories not there:

> ls -lrth /mnt/remote/Media/TV |wc -l
> 120

**Comment by darthShadow** [comment]
@H1XYZ Can you add the logs from the mount command too?

**Comment by ncw** [comment]
> @H1XYZ Can you add the logs from the mount too while you ran this command?

That would be super helpful :-)

**Comment by H1XYZ** [comment]
> > @H1XYZ Can you add the logs from the mount too while you ran this command?
> 
> That would be super helpful :-)

Command:

> /usr/bin/rclone mount --config=/home/seed/.config/rclone/rclone.conf --allow-other --rc --rc-addr=localhost:5572 --fast-list --drive-skip-gdocs --vfs-read-chunk-size=64M --vfs-read-chunk-size-limit=2048M --buffer-size=64M --max-read-ahead=256M --poll-interval=1m --dir-cache-time=168h --timeout=10m --transfers=16 --checkers=8 --drive-chunk-size=64M --fuse-flag=sync_read --fuse-flag=auto_cache --umask=002 --log-file=/home/seed/logs/rclone_mount.log -vvvvv google: /mnt/remote

then ran:

> /usr/bin/rclone \
>   --config /home/seed/.config/rclone/rclone.conf \
>   --timeout=1h \
>   -vvvvv \
>   --rc-addr=localhost:5572 \
>   rc vfs/refresh dir=Media recursive=true

Result:
> ➜  ~ ls -lrth /mnt/remote/Media/TV |wc -l
> 332
> ➜  ~ sudo systemctl start rclone_vfs_primer.service
> ➜  ~ ls -lrth /mnt/remote/Media/TV |wc -l
> 93

Log:
https://termbin.com/lhl3

**Comment by darthShadow** [comment]
Would it be possible for you to get the logs without `--fast-list` and the same tps limits?

**Comment by ncw** [comment]
I have found a problem, not 100% sure it is the problem you are seeing though.

If a directory has multiple parents it will only appear in one place when used with `--fast-list`

```
$ rclone lsf -R -Fip drive:test/move
17ZTnUFqOP01KciVydeXVoZOfT4kW1lbo;Folder/
1DFPx0tuWMzrCfN_nmDw6LqH7MaDr_3mU;srcDir/
1LRGw-qQvghayrZVo8Z_iaeiCiOMaq1Eo;srcDir2/
1QS-g4CkUIEp0_XRMFwNctWhMuxV_-5eD;Folder/srcDir/
1LRGw-qQvghayrZVo8Z_iaeiCiOMaq1Eo;Folder/srcDir2/
14mpPQ6wxX6KKYzpBIhHBPG81DAQdmAHt;Folder/srcDir3/
```

Note how the `srcDir2` directory is missing from `Folders` but present twice in the root.

```
$ rclone --fast-list lsf -R -Fip drive:test/move
17ZTnUFqOP01KciVydeXVoZOfT4kW1lbo;Folder/
1DFPx0tuWMzrCfN_nmDw6LqH7MaDr_3mU;srcDir/
1LRGw-qQvghayrZVo8Z_iaeiCiOMaq1Eo;srcDir2/
1LRGw-qQvghayrZVo8Z_iaeiCiOMaq1Eo;srcDir2/
1QS-g4CkUIEp0_XRMFwNctWhMuxV_-5eD;Folder/srcDir/
14mpPQ6wxX6KKYzpBIhHBPG81DAQdmAHt;Folder/srcDir3/
```

This replicates your problem, but I don't think it is your problem since you reported that 

```
rclone lsf -R gdrive:troublesomedir | sort > test0
rclone lsf --fast-list -R gdrive:troublesomedir | sort > test1
rclone lsf --fast-list -R gdrive:troublesomedir | sort > test2
diff -u test0 test1
diff -u test1 test2
```

Reported no differences.

You can test for the above problem like this:

```
rclone --fast-list lsf -R -Fip google: | sort | uniq -d
```

If it prints anything, then you have the above problem.

**Comment by H1XYZ** [comment]
@ncw 

I ran this:

> rclone --fast-list lsf -R -Fip google:Media/TV | sort | uniq -d

Which had no result (meaning it's ok)

Then ran:

> rclone --fast-list lsf -R -Fip google:Media/ | sort | uniq -d

and got a bunch of folders listed similar to your example posted above

**Comment by B4dM4n** [comment]
I found the issue in the drive `ListR` implementation and created a fix for it. 

Once the PR is merged `ListR` will correctly handle items with multiple parents, at the cost of scanning those folders once per parent if needed.

In general the handling of items with multiple parents might not behave as expected in all circumstances. e.g. deleting a item(file or folder) with multiple parents will remove it from all parents.

**Comment by EnorMOZ** [comment]
Once a build/release for this issue is in place we can test it out. Thanks again for the support in this !!

**Comment by ncw** [comment]
I pushed @B4dM4n 's branch to the main repo to make a beta for you to try

https://beta.rclone.org/branch/v1.45-185-g177c274f-drive-multi-parents-beta/ (uploaded in 15-30 mins)

**Comment by H1XYZ** [comment]
> I pushed @B4dM4n 's branch to the main repo to make a beta for you to try
> 
> https://beta.rclone.org/branch/v1.45-185-g177c274f-drive-multi-parents-beta/ (uploaded in 15-30 mins)

Confirming that this has resolved the issue!

Pre:

> ➜  ~ ls -lrth /mnt/remote/Media/TV |wc -l
> 332
> 

Post:

> ➜  ~ ls -lrth /mnt/remote/Media/TV |wc -l  
> 332

Thank you to everyone involved in figuring this one out 👍

**Comment by ncw** [comment]
> Confirming that this has resolved the issue!

Excellent news!

> Thank you to everyone involved in figuring this one out

I think there were quite a few contributors :clap: 

We've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.46
